src/daft-plan/src/builder.rs

```diff
@@ -217,53 +190,233 @@ impl LogicalPlanBuilder {
         &self,
         root_dir: &str,
         file_format: FileFormat,
-        partition_cols: Option<Vec<PyExpr>>,
+        partition_cols: Option<Vec<Expr>>,
         compression: Option<String>,
-    ) -> PyResult<LogicalPlanBuilder> {
-        let part_cols =
-            partition_cols.map(|cols| cols.iter().map(|e| e.clone().into()).collect::<Vec<Expr>>());
+    ) -> DaftResult<Self> {
         let sink_info = SinkInfo::OutputFileInfo(OutputFileInfo::new(
             root_dir.into(),
             file_format,
-            part_cols,
+            partition_cols,
             compression,
         ));
         let fields = vec![Field::new("file_paths", DataType::Utf8)];
         let logical_plan: LogicalPlan = ops::Sink::new(
+            self.plan.clone(),
             Schema::new(fields)?.into(),
             sink_info.into(),
-            self.plan.clone(),
         )
         .into();
         Ok(logical_plan.into())
     }
 
-    pub fn optimize(&self) -> PyResult<LogicalPlanBuilder> {
-        let optimizer = Optimizer::new(Default::default());
-        let new_plan = optimizer.optimize(self.plan.clone(), |_, _, _, _, _| {})?;
-        Ok(Self::new(new_plan))
+    pub fn build(&self) -> Arc<LogicalPlan> {
+        self.plan.clone()
+    }
```

Comment:
I was initially going to make this a consuming op (taking `self` rather than `&self` and directly returning `self.plan`) but I ran into some pyo3 issues with `PyLogicalPlanBuilder`, which can't be consumed by one of its methods. Going to try a few options to achieve this in a future PR while hopefully keep things clean.

Reply:
Right, I started messing around with a few different patterns for optimization, where even though we consume the plan we replace it with an optimized plan, but for computation, which would actually consume the plan without replacing it, I don't think we could have those semantics.